File: Projekti/Hibernate-Benchmark/src/main/java/com/example/HibernateBenchmark/repository/UserRepository.java
package com.example.HibernateBenchmark.repository;

import com.example.HibernateBenchmark.dto.UserPostCount;
import com.example.HibernateBenchmark.dto.UserProfile;
import com.example.HibernateBenchmark.model.User;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.time.LocalDate;
import java.util.List;

public interface UserRepository extends JpaRepository<User, Integer> {

    List<User> findByDateOfBirthAfter(LocalDate year);

    List<User> findByUserIdGreaterThan(Integer id);

    @Query(value = "SELECT u FROM User u WHERE u.dateOfBirth > :year")
    List<User> getUsersBornAfter(@Param("year") LocalDate year);

    @Query( "SELECT new com.example.HibernateBenchmark.dto.UserPostCount(u.userId, u.email, size(u.posts))" +
            "FROM User u")
    List<UserPostCount> findUsersWithPostCount();

    @Query("SELECT u.userId, u.email, p.bio FROM Profile p JOIN p.user u WHERE p.bio LIKE %:keyword%")
    List<UserProfile> findUsersByBioKeyword(@Param("keyword") String keyword);

    @Query("SELECT new com.example.HibernateBenchmark.dto.UserPostCount(u.userId, u.email, size(u.posts))" +
            " FROM User u JOIN u.posts p GROUP BY u.userId ORDER BY COUNT(p.postId) DESC LIMIT 5")
    List<Object[]> findTopPosters();

}

File: Projekti/EntityFramework-Benchmark/Repository/UserRepository.cs
using EntityFramework_Benchmark.Dto;
using EntityFramework_Benchmark.Model;
using EntityFramework_Benchmark.Repository.IRepository;
using EntityFramework_Benchmark.Settings;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace EntityFramework_Benchmark.Repository
{
    public class UserRepository : IUserRepository
    {
        public BenchmarkDbContext Context { get; set; }

        public UserRepository() 
        {
            Context = new BenchmarkDbContext();
        }

        public UserRepository(BenchmarkDbContext context) 
        {
            Context = context;
        }

        public List<User> GetUsersBornAfter(DateTime year)
        {
            return Context.Users.Where(x => x.DateOfBirth >= year).ToList();
        }

        public List<UserPostCount> GetNumberOfPostsForEachUser()
        {
            return Context.Users.Include(x => x.Posts).Select(x => new UserPostCount { Id = x.UserId, Email = x.Email, Count = x.Posts.Count }).ToList();
        }

        public List<UserPostCount> GetTopFiveUsersByPostCount()
        {
            return Context.Users
                .Include(x => x.Posts)
                .Select(u => new UserPostCount { Id = u.UserId, Email = u.Email, Count = u.Posts.Count() })
                .OrderByDescending(x => x.Count)
                .Take(5)
                .AsSingleQuery()
                .ToList();      
        }

        public List<UserProfile> GetUsersWithKeywordInBio(string keyword)
        {
            return Context.Users
                    .Include(x => x.Profiles)
                    .Where(u => u.Profiles.First().Bio.Contains(keyword))
                    .Select(u => new UserProfile(u.UserId, u.Email, u.Profiles.First().Bio))
                    .ToList();
        }

        public void InsertUser(User user)
        {
            Context.Users.Add(user);
            Context.SaveChanges();
        }
    }
}
